platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/ValueState.java

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import org.gradle.api.Action;
+import org.gradle.internal.Cast;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.logging.text.TreeFormatter;
+import org.gradle.internal.state.ModelObject;
+
+import javax.annotation.Nullable;
+import java.util.function.Supplier;
+
+/**
+ * Manages values that are finalizable and support conventions.
+ *
+ * @param <S>
+ */
+public abstract class ValueState<S> {
+    private static final ValueState<Object> FINALIZED_VALUE = new FinalizedValue<>();
+
+    public static <S> ValueState<S> newState(PropertyHost host) {
+        return new ValueState.NonFinalizedValue<>(host);
+    }
+
+    public abstract boolean shouldFinalize(Supplier<DisplayName> displayName, @Nullable ModelObject producer);
+
+    public abstract ValueState<S> finalState();
+
+    abstract void setConvention(S convention);
+
+    public abstract void disallowChanges();
+
+    public abstract void finalizeOnNextGet();
+
+    public abstract void disallowUnsafeRead();
+
+    public abstract S explicitValue(S value);
+
+    public abstract S explicitValue(S value, S defaultValue);
+
+    public abstract S applyConvention(S value, S convention);
+
+    /**
+     * Marks this value state as being non-explicit. Returns the convention, if any.
+     */
+    public abstract S implicitValue();
+
+    public abstract boolean maybeFinalizeOnRead(Supplier<DisplayName> displayName, @Nullable ModelObject producer, ValueSupplier.ValueConsumer consumer);
+
+    public abstract void beforeMutate(Supplier<DisplayName> displayName);
+
+    public abstract ValueSupplier.ValueConsumer forUpstream(ValueSupplier.ValueConsumer consumer);
+
+    public boolean isFinalized() {
+        return this == FINALIZED_VALUE;
+    }
+
+    /**
+     * Is this state final or on its way for being finalized?
+     */
+    public abstract boolean isFinalizing();
+
+    public void finalizeOnReadIfNeeded(Supplier<DisplayName> getDisplayName, ModelObject effectiveProducer, ValueSupplier.ValueConsumer consumer, Action<ValueSupplier.ValueConsumer> finalizeNow) {
```

Comment:
This replaces the original implementation of `AbstractProperty#beforeRead()`:

https://github.com/gradle/gradle/pull/27134/files#diff-4dfacf9a118d2e6d76280d86e28279d8213db70d47ec6efbf630d231817c6640L238